File: algorithms/153.Find_Minimum_in_Rotated_Sorted_Array/Find_Minimum_in_Rotated_Sorted_Array.py
class Solution:
    # @param num: a rotated sorted array
    # @return: the minimum number in the array
    def findMin(self, nums):
        if len(nums) == 0:
            return 0
            
        start, end = 0, len(nums) - 1
        target = nums[-1]
        while start + 1 < end:
            mid = (start + end) / 2
            if nums[mid] <= target:
                end = mid
            else:
                start = mid
        return min(nums[start], nums[end])
    class Solution:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        low, high = 0, len(num)
           
        while low < high - 1 and  num[low] >= num[high - 1]:
            mid = low + (high - low) / 2
                
            if num[mid] > num[low]:
                low = mid + 1
            elif num[mid] < num[low]:
                if mid == high - 1:
                    return num[mid]
                else:
                    high = mid + 1
            else:
                return num[mid]

        return num[low]

class Solution2:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        low, high = 0, len(num) - 1
           
        while low < high and num[low] >= num[high]:
            mid = low + (high - low) / 2
            
            if num[mid] >= num[low]:
                low = mid + 1
            else:
                high = mid

        return num[low]

if __name__ == "__main__":
    print Solution().findMin([1])
    print Solution().findMin([1, 2])
    print Solution().findMin([2, 1])
    print Solution().findMin([3, 1, 2])
    print Solution().findMin([2, 3, 1])

File: algorithms/153.Find_Minimum_in_Rotated_Sorted_Array/Find_Minimum_in_Rotated_Sorted_Array.cpp
<<<<<<< HEAD
#include <bits/stdc++.h>

class Solution {
public:
    int findMin(vector<int> &num) {
        int left = 0, right = num.size() - 1;
        int mid;
        while (left <= right) {
            mid = (left + right) / 2;
            if (num[left] <= num[mid] && num[mid] <= num[right]) {
                return num[left];
            } else if (num[left] <= num[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return num[0];
    }
};

int main() {
    vector<int> num;
    num.push_back(3);
    num.push_back(1);
    num.push_back(2);
    Solution s;
    cout << s.findMin(num) << endl;
    return 0;
}

=======
#include <bits/stdc++.h>

class Solution {
public:
    int findMin(vector<int> &num) {
        int left = 0, right = num.size() - 1;
        int mid;
        while (left <= right) {
            mid = (left + right) / 2;
            if (num[left] <= num[mid] && num[mid] <= num[right]) {
                return num[left];
            } else if (num[left] <= num[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return num[0];
    }
};

int main() {
    vector<int> num;
    num.push_back(3);
    num.push_back(1);
    num.push_back(2);
    Solution s;
    cout << s.findMin(num) << endl;
    return 0;
}

>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
